Add support for bootstrap containers via API settings (#1387)
bottlerocket-os/bottlerocket: README.md

Be careful, and make sure you have a similar low-level use case before reaching for host containers.

#### Bootstrap containers settings

> * `settings.bootstrap-containers.<name>.source`: The URI of the container

Reviewer: Just curious - is this a single container we can run, or can we set up several? The plural on `bootstrap-containers` is confusing me a bit.

Author: Hi @diranged, thanks for the feedback, I'll try to make this bit clearer. To answer your question, you could setup several bootstrap containers, but they should be independent from each other since we don't implement a mechanism to run them in a specific order.

Reviewer: Ok great. Also, is there support for these containers being hosted in a private registry?

Author: Since these host containers, you can use public/private ECR registries, as well as public registries.

* `settings.bootstrap-containers.<name>.mode`: The mode of the container, it could be one of `off`, `once` or `always`
* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data

Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ecs-agent, kubelet and docker start.

> This type of containers are very similar to normal host containers, they come with persistent storage and with optional user data. Unlike normal host containers, bootstrap containers can't be threated as `superpowered` containers. However, these containers run as privileged and have access to the underlying root filesystem on `/.bottlerocket/rootfs`.

Reviewer: s/threated/treated/

Author: Nice catch! Thanks

Bootstrap containers have three different modes:

> * `always`: with this setting, the container is executed in all boots.

Reviewer 2: s/in all boots/on every boot

Author: I like that better, thanks! :+1:

> * `off`: blocks the execution of the container

Reviewer 2: It might be clearer to say "the container will not run"

Author: :+1:

Be careful, and make sure you have a similar low-level use case before reaching for host containers.

#### Bootstrap containers settings

> * `settings.bootstrap-containers.<name>.source`: the URI of the container

Reviewer 3: nit:
```suggestion
* `settings.bootstrap-containers.<name>.source`: the image for the container
```

Images are not specified as URIs since they lack a scheme.

Author: Fixed

> * `settings.bootstrap-containers.<name>.mode`: the mode of the container, it could be one of `off`, `once` or `always`

Reviewer 3: What's the difference between `once` and `always`?  ~I assume that it's "first boot" vs. "every boot" but I don't think that's clear here.~

This is explained below, maybe add a pointer to it.

```suggestion
* `settings.bootstrap-containers.<name>.mode`: the mode of the container, it could be one of `off`, `once` or `always`.  See below for a description of modes.
```

Author: Fixed

* `settings.bootstrap-containers.<name>.essential`: whether or not the container should fail the boot process, default to `false`
* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data

> Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.

Reviewer 4: I'd rather not call them host-containers if we're making a distinction between the two :)

```suggestion
Bootstrap containers, similar to host containers, are containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.
```

Author: I wanted to make sure users understand that bootstrap containers have the same caveats as host containers, without repeating them, should I repeat the caveats?

Reviewer 4: Seems worthwhile, maybe point to the docs saying "and the same caveats apply"

This type of container is very similar to normal host containers, they come with persistent storage and with optional user data.
Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
However, these containers have access to the underlying root filesystem on `/.bottlerocket/rootfs`.

> Bootstrap containers are started once the `configured` systemd's target is reached, so they are not executed in a deterministic order.

Reviewer 4: This might be slightly pedantic, but I think this more accurately clarifies why the dependency ordering can't provide deterministic ordering.

```suggestion
Bootstrap containers are set up to run after the systemd `configured.target` unit is active.
The containers' systemd unit depends on this target (and not on any of the bootstrap containers' peers) which means that bootstrap containers will not execute in a deterministic order.
```

Author: Fixed

The boot process will "wait" for as long as the bootstrap containers run.
Bootstrap containers configured with `essential=true` will stop the boot process if they exit code is a non-zero value.

Bootstrap containers have three different modes:

* `always`: with this setting, the container is executed on every boot.
* `off`: the container won't run

> * `once`: with this setting, the container only runs on the first boot. Upon completion, the container marks itself as `off`.

Reviewer 3: ```suggestion
* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is changed to `off`.
```

Be careful, and make sure you have a similar low-level use case before reaching for host containers.

#### Bootstrap containers settings
* `settings.bootstrap-containers.<name>.source`: The URI of the container
* `settings.bootstrap-containers.<name>.mode`: The mode of the container, it could be one of `off`, `once` or `always`
* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data

> Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ecs-agent, kubelet and docker start.

Reviewer 3: nit: "ecs-agent, kubelet, and docker start" to use the Oxford comma.

second nit: those are the names of the services in systemd, which is one correct way to do this.  The other way would be to use the names of the software which would be "ECS Agent", "Kubelet", and "Docker".  Is it your intention to use the systemd service names?

Author: No, my intention was to used the services names. I already updated them.

> This type of containers are very similar to normal host containers, they come with persistent storage and with optional user data.

Reviewer 5: nit

```suggestion
These types of containers are very similar to normal host containers, they come with persistent storage and with optional user data.
```

or

```
This type of container is very similar
```

or

```
Bootstrap containers are very similar
```

Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
However, these containers run as privileged and have access to the underlying root filesystem on `/.bottlerocket/rootfs`.

> Bootstrap containers are started once the `configured` systemd's target is reached, so they are not executed in a deterministic order.

Reviewer 5: I wonder if deterministic order would be 'nice-to-have' feature. One thing that's not clear currently in the documentation is whether the bootstrap containers are fired off in parallel or one-after-another.

Author: I updated the documentation, and I mentioned that bootstrap containers run in parallel. Regarding the deterministic order, I think use cases that require specific order should belong to one container instead of splitting work and orchestrating containers. Do you have an specific use case in mind?

> The boot process will "wait" for as long as the bootstrap containers run, and if any of them fail, the boot process will stop.

Reviewer 3: Might be worth adding a second sentence here: "Bootstrap containers must exit with an exit code of 0 before the boot process can continue."

Author: Done :+1:

Bootstrap containers have three different modes:

* `always`: with this setting, the container is executed on every boot.
* `off`: the container won't run
* `once`: with this setting, the container only runs on the first boot. Upon completion, the container marks itself as `off`.
If you need to execute the containers after the first boot, you can mark it as `once` and reboot.

Here's an example of adding a bootstrap container with API calls:

```
apiclient -u /settings -X PATCH -d '{"bootstrap-containers": {"bootstrap": {"source": "MY-CONTAINER-URI", "mode": "once"}}}'

> apiclient -u /tx/commit_and_apply -X POST

Reviewer 3: Can this example be written with `apiclient set`?  Or are multiple fields required here such that the simplified `set` mode is insufficient?

Author: I can write it with  `apiclient set --json` and specify the settings as a json blob or write multiple lines with `apiclient set bootstrap-containers.blah.*`

Author: Done :+1:

Be careful, and make sure you have a similar low-level use case before reaching for host containers.

#### Bootstrap containers settings
* `settings.bootstrap-containers.<name>.source`: the image for the container
* `settings.bootstrap-containers.<name>.mode`: the mode of the container, it could be one of `off`, `once` or `always`. See below for a description of modes.
* `settings.bootstrap-containers.<name>.essential`: whether or not the container should fail the boot process, default to `false`
* `settings.bootstrap-containers.<name>.user-data`: field with arbitrary base64-encoded data

Bootstrap containers are host containers that can be used to "bootstrap" the host before services like ECS Agent, Kubernetes, and Docker start.

> This type of container is very similar to normal host containers, they come with persistent storage and with optional user data.

Reviewer 6: ```suggestion
These container are very similar to normal host containers.
They come with persistent storage and with optional user data.
```

Author: Fixed

Unlike normal host containers, bootstrap containers can't be treated as `superpowered` containers.
However, these containers have access to the underlying root filesystem on `/.bottlerocket/rootfs`.
Bootstrap containers are set up to run after the systemd `configured.target` unit is active.
The containers' systemd unit depends on this target (and not on any of the bootstrap containers' peers) which means that bootstrap containers will not execute in a deterministic order
The boot process will "wait" for as long as the bootstrap containers run.
Bootstrap containers configured with `essential=true` will stop the boot process if they exit code is a non-zero value.

Bootstrap containers have three different modes:

* `always`: with this setting, the container is executed on every boot.
* `off`: the container won't run

> * `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is chaged to `off`.

Reviewer 6: ```suggestion
* `once`: with this setting, the container only runs on the first boot where the container is defined. Upon completion, the mode is changed to `off`.
```

Author: Fixed

> If you need to execute the containers after the first boot, you can mark it as `once` and reboot.

Reviewer 6: nit: not sure we need this note, it seems like the result of this might not be very well-defined.

Here's an example of adding a bootstrap container with API calls:

```
apiclient set \
   bootstrap-containers.bootstrap.source=MY-CONTAINER-URI \
   bootstrap-containers.bootstrap.mode=once \
   bootstrap-containers.bootstrap.essential=false
```

Here's the same example, but with the settings you'd add to user data:

```
[settings.bootstrap-containers.bootstrap]
source = "MY-CONTAINER-URI"
mode = "once"

> essential = false

Reviewer 6: Should we suggest `essential = true` as the default? It's more in keeping with the behavior of the overall system.

Author: Done
